File: Example/Python/MySampleTextPlugin.py
# 必须: 导入 C++ 绑定的模块
import gpp_plugin_api as gpp
from pathlib import Path

# Sentence 的声明详见 Example/Lua/MySampleTextPlugin.lua

# 全局变量，由 C++ 注入
# python 中的 logger 和 toknizeFunc 不在 utils 中而是在当前模块的全局变量中
logger = None

# sourceLang_useTokenizer = True
# sourceLang_tokenizerBackend = "MeCab"
# sourceLang_meCabDictDir = "..."
# sourceLang_spaCyModelName = "ja_core_news_sm"
# targetLang_spaCyModelName = "..."
# targetLang_stanzaLang = "..."
# sourceLang_tokenizeFunc = None
# ...

targetLang_useTokenizer = True
targetLang_tokenizerBackend = "pkuseg"
# 如果使用提供的分词器则必须先定义 tokenizeFunc
targetLang_tokenizeFunc = None


def checkConditionForRetranslKeysFunc(se: gpp.Sentence) -> bool:
    if se.index == 278:
        logger.info("RetransFromPython")
    return False

def checkConditionForSkipProblemsFunc(se: gpp.Sentence) -> bool:
    if se.index == 278:
        # retranslKey 中检查的 se 是一个副本，只负责检查，怎么设置都不会影响到要翻译和输出的部分
        # 但 skipProblems 的检查提供的是要输出的 se 的引用
        current_problem = ""
        for i, s in enumerate(se.problems):
            # 正则检查通过后，执行 Condition 验证时，当前正在检查的 problem 会拥有前缀 'Current problem:'
            # PS: 不要在此检查函数中的任何地方以任何方式修改 problems 对象本身，可能导致 C++侧 出现异常
            # 但可以通过 problems_set_by_index 修改元素内容曲线救国
            if s.startswith("Current problem:"):
                current_problem = s[16:]
            else:
                success = se.problems_set_by_index(i, "My new problem")
                if success:
                    se.other_info |= { f"pySuccess{i}": "New problem suc" }
                else:
                    se.other_info |= { f"pyFail{i}": "New problem fail" }

        se.other_info |= { "pythonCheck": "The 278th" }
        se.other_info |= { "pyCp": current_problem }
        logger.info("SkipProblemsFromPython")
    return False

def init(project_dir: Path):
    """
    插件初始化函数，由 C++ 调用一次。
    """
    logger.info(f"MySampleTextPluginFromLua 初始化成功，projectDir: {project_dir}")


def postRun(se: gpp.Sentence):
    """
    处理每个句子的主函数。
    se 是一个 C++ Sentence 对象的代理。
    """
    if se.translated_preview == "久远紧锁眉头，脸上写着「骗人的吧，喂」。":
        wordpos_vec, entity_vec = targetLang_tokenizeFunc(se.translated_preview)
        # 其它(虽然没几个)工具函数依然在 utils 里
        tokens = gpp.utils.splitIntoTokens(wordpos_vec, se.translated_preview)
        parts = [f"[{value[0]}]" for value in wordpos_vec]
        result = "#".join(parts)
        # 所有容器均为 copy，直接对其进行 insert，append 等是无效的
        se.other_info |= { "tokens_python": result }

# def unload():
#     pass

# preRun, run, postRun, unload 均可选择性定义
# 对于译前插件，如果定义 preRun/run 函数，则分别在对应阶段调用
# 对于译后插件，如果定义 postRun/run 函数，则分别在对应阶段调用

File: Example/Lua/MySampleTextPlugin.lua
--[[
enum class NameType
{
    None, Single, Multiple
};

struct Sentence {
    int index;
    std::string name;
    std::vector<std::string> names;
    std::string name_preview;
    std::vector<std::string> names_preview;
    std::string original_text;
    std::string pre_processed_text;
    std::string pre_translated_text;
    std::vector<std::string> problems;
    std::string translated_by;
    std::string translated_preview;
    std::string originalLinebreak;
    std::map<std::string, std::string> other_info;

    NameType nameType = NameType::None;
    Sentence* prev = nullptr;
    Sentence* next = nullptr;
    bool complete = false;
    bool notAnalyzeProblem = false;

    std::optional<std::string> problems_get_by_index(int index);
    bool problems_set_by_index(int index, const std::string& problem);
};

utils里的工具函数的签名详见 LuaMnager::registerCustomTypes 中绑定 utils 库的部分
]]

sourceLang_useTokenizer = false
targetLang_useTokenizer = true
--sourceLang_tokenizerBackend = "MeCab"
targetLang_tokenizerBackend = "pkuseg"
--sourceLang_meCabDictDir = "..."
--targetLang_spaCyModelName = "..."
--targetLang_stanzaLang = "..."
--...

function checkConditionForRetranslKeysFunc(sentence)
    if sentence.index == 278 then 
        utils.logger:info("RetransFromLua")
    end
    return false
end

function checkConditionForSkipProblemsFunc(sentence)
    if sentence.index == 278 then 
        -- Example/Python/MySampleTextPlugin 中有对此函数的进一步介绍
        current_problem = ""
        for i, s in ipairs(sentence.problems) do
            if string.find(s, "^Current problem:") then 
                current_problem = string.sub(s, 17)
            else
                local success = sentence:problems_set_by_index(i-1, "My new problem")
                if success then
                    sentence.other_info["luaSuccess"] = "New problem suc"
                else
                    sentence.other_info["luaFail"] = "New problem fail"
                end
            end
        end
        local other_info_map = sentence.other_info
        other_info_map["luaCheck"] = "The 278th"
        other_info_map["luaCp"] = current_problem
        sentence.other_info = other_info_map
    end
    return false
end

function dPostRun(sentence)
    -- 为了能使用 pairs 遍历，map 返回类型都是副本，如果想修改，必须 mapVar = copy
    -- 但非递归的 vector 类型可使用 sol2 提供的 container 通用函数进行引用修改，具体函数及其作用详见
    -- https://github.com/ThePhD/sol2/blob/main/documentation/source/containers.rst 的 container operations 部分
    local other_info_map = sentence.other_info
    if sentence.index == 0 then
        other_info_map["MySampleKeyForFirstSentence"] = "MySampleValueForFirstSentence"
        sentence.other_info = other_info_map
    elseif sentence.index == 278 then
        sentence.problems:add("278FlagProblem")
    end
    if sentence.next == nil and sentence.prev ~= nil and sentence.prev.nameType == NameType.None then
        sentence.problems:add("MySampleProblemForLastSentence")
        sentence.notAnalyzeProblem = true
    end
    if sentence.translated_preview == "久远紧锁眉头，脸上写着「骗人的吧，喂」。" then
        -- utils, json, toml 等无实例类的纯工具表直接使用 .func(...) 调用即可
        -- logger，translator 等有具体实例的对象需要用 :func(...) 或 .func(self, ...) 调用
        local wordpos_vec, entity_vec = utils.targetLang_tokenizeFunc(sentence.translated_preview)
        local parts = {}
        for i, value in ipairs(wordpos_vec) do
            parts[i] = "[" .. value[1] .. "]"
        end
        local result = table.concat(parts, "#")
        other_info_map["tokens_lua"] = result
        utils.logger.info(utils.logger, string.format("测试 logger ，当前 index: %d", sentence.index))
        sentence.other_info = other_info_map
    end
end

function init(projectDir)
    utils.logger:info(string.format("MySampleTextPluginFromLua 初始化成功，projectDir: '%s'", projectDir.value))
end

-- function unload()
--     -- ctrl + /
--     -- return
--     -- utils.logger:info("MySampleTextPluginFromLua unloads")
--     -- local cmd = [[D:\GALGAME\linshi\天冥のコンキスタ\WORK\#toSjis.bat]]
--     -- local cmd_acp = utils.ascii2Ascii(cmd, 65001, 0)
--     -- local handle = io.popen(cmd_acp, 'r')
--     -- if handle then
--     --     local output = handle:read("*a")
--     --     handle:close()
--     --     utils.logger:info("--- BAT 脚本的输出内容 开始 ---\n" .. output .. "\n--- BAT 脚本的输出内容 结束 ---")
--     -- else
--     --     utils.logger:error("无法执行 BAT")
--     -- end
-- end

-- dPreRun, preRun, postRun, dPostRun, unload 均可选择性定义，如果定义则分别在对应阶段调用
